File: Leetcode/238. Product of Array Except Self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        zeros = 0
        
        for num in nums:
            if num == 0:
                zeros += 1
                if zeros == 2:
                    return [0] * len(nums)
            else:
                product *= num
        
        if zeros:
            return [0 if num != 0 else product for num in nums]
        else:
            return [product // num for num in nums]

File: Leetcode/238. Product of Array Except Self/solution.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut product = 1;
        let mut zeros = 0;

        for &num in &nums {
            if num == 0 {
                zeros += 1;
                if zeros == 2 {
                    return vec![0; nums.len()];
                }
            } else {
                product *= num;
            }
        }

        if zeros > 0 {
            nums.iter().map(|&num| if num != 0 { 0 } else { product }).collect()
        } else {
            nums.iter().map(|&num| product / num).collect()
        }
    }
}

